src/runtime/runtime.cs

```diff
@@ -748,6 +748,22 @@ internal static IntPtr SelfIncRef(IntPtr op)
             return op;
         }
 
+        /// <remark>
+        /// We need this method because BorrowedReference can be implicitly casted to IntPtr.
+        /// </remark>
+        internal static void XDecref(BorrowedReference op)
+        {
+            throw new InvalidOperationException("Cannot DecRef a borrowed reference.");
+        }
+
+        /// <remark>
+        /// We need this method because NewReference can be implicitly casted to IntPtr.
+        /// </remark>
+        internal static void XDecref(NewReference op)
+        {
+            op.Dispose();
+        }
+
```

Comment:
No implicit casts. The types were created specifically to avoid freely mixing different kinds of ownership for `PyObject*`.